executor-scripts/linux/vrrp

```diff
@@ -0,0 +1,358 @@
+#
+# Move the crc32sum tool to internal applet of ifupdown-ng and change the VRRP 
+# executor to use the new applet instead external crc32sum tool.
+# The new applet is called ifcrc32sum.
+#
+# Sun, 03 Jan 2024 19:00:26 +0300
+#  -- EasyNetDev <devel@easynet.dev>
+#
+# This executor will do the configuration for MACVLAN virtual interfaces.
+#
+# Each MACVLAN interface must be controlled by a VRRP daemon, like FRRouting.
+#
+# Options needed for VRRP executro to be able to create the MACVLAN interface are:
+#
+# IFACE                Main interface name
+# IF_VRRP_ID           VRID list to create the interfaces only
+# IF_VRRP_ADDRESS      IP list to create the interfaces and add IPs (Optional)
+#
+# Interface naming would be in format vXXXXXXXX-YY-V
+# Where
+#  ->  XXXXXXXX is CRC32 value calculated for main interface name
+#  ->  YY	is HEX value of the VRRP ID
+#  ->  V	is IP version 4 or 6
+#
+# This keeps the a fix VRRP interface name length to 14 characters (maximum are 15 characters in Linux)
+# and the part vXXXXXXXX-YY is the same for both IP protocols.
+# Also it keeps the same names after reboots if the main interface name is not changed or the CRC32 tool is not changed.
+#
+# This script is written in shell and is not using any Bash built-in functions.
+#
+
+[ -n "$VERBOSE" ] && set -x
+
+# Global variables for CRC32 tool
+crc32_cmd=""
+crc32_tool=""
+
+################################################################################
+#                        VRRP CRC32 tool functions                             #
+################################################################################
+
+# In case the crc32_tool is in auto, search for ifcrc32sum then for cksum
+find_crc32_tools_in_auto() {
+    # In test mode, we need to MOCK aka echo output.
+    if [ -n "${MOCK}" -a -z ${crc32_cmd} ]; then
+	crc32_tool="MOCK"
+	return
+    fi
+    # Search for ifcrc32sum
+    crc32_cmd=$(command -v ifcrc32sum)
+    if [ -n "${crc32_cmd}" ]; then
+	crc32_tool="ifcrc32sum"
+	return
+    fi
+    # Search for cksum
+    crc32_cmd=$(command -v cksum)
+    if [ -n "${crc32_cmd}" ]; then
+	crc32_tool="cksum"
+	return
+    fi
+    # If none of them are found, print an error message
+    if [ -z "${crc32_cmd}" ]; then
+	echo "Couldn't find crc32sum or cksum tools! Please build ifupdown-ng with ifcrc32sum support or install coreutils package for cksum!"
+	exit 0
+    fi
+}
+
+# Compute CRC32 value using the selected CRC32 tool
+compute_crc32_hex() {
+    case "$crc32_tool" in
+	ifcrc32sum)
+	    # ifcrc32sum has an option -x to directly outputs CRC32 value in hex.
+	    $crc32_tool -x "$1"
+	    ;;
+	cksum)
+	    # cksum has 2 fields: CRC32 value and total length of the string. We need to take first value and convert to hex.
+	    printf "%08x" $(echo -n "$1" | ${crc32_cmd} | cut -f1 -d" ")
+	    ;;
+	MOCK)
+	    # This is CRC32 hex value calculated for "eth0" string. Both ifcrc32sum and cksum are giving the same value.
```

Comment:
Err, what?  We should definitely be testing that ifcrc32sum is generating the correct CRC32, even when running under tests.

Reply:
Please ask your colleague that comment to my previous PR. Now I have no clue why you asking me to add exactly what I removed from previous PR.

Reply:
I am confused, too :) Needless to say though, it would be good to have test coverage of ifcrc32sum with known-correct outputs.

Reply:
My comment was that if you are going to mock a value for tests, then it needs to be mocked.  Testing the tool should be done with a separate test case, ideally using multiple fixtures.  Additionally, the original PR could call whatever the system `cksum` binary was for testing purposes, which means if something was broken with it, the ifupdown-ng test would fail.

While I'm here: my suggestion that users should be allowed to sub in their own crc32_cmd was never applied, either.  There are multiple types of systems that could have an accelerated crc32 command and it may not be named the same as the standard command.

Reply:
Hi @awilfox

If the user has it own tool, how can we allow for their tool?
Maybe you are referring to an option in ifupdown.conf where the user can set arbitrary CRC32?

My executor has already 2 choises. To allow the user for their tool, I don't see any ways than setting a specific option in configuration file.

Reply:
> I am confused, too :) Needless to say though, it would be good to have test coverage of ifcrc32sum with known-correct outputs.

For cksum and ifcrc32sum we know for sure that should output the same value. So I can do a test using these tools. But just for cksum and ifcrc32sum.